lib/Dialect/Arc/Transforms/ArcCanonicalizer.cpp

```diff
@@ -0,0 +1,299 @@
+/// up-to-date no matter whether the pattern succeeds of fails.
+template <typename SourceOp>
+class SymOpRewritePattern : public OpRewritePattern<SourceOp> {
+public:
+  SymOpRewritePattern(MLIRContext *ctxt, SymbolHandler &symbolCache,
+                      Namespace &names, mlir::PatternBenefit benefit = 1,
+                      ArrayRef<StringRef> generatedNames = {})
+      : OpRewritePattern<SourceOp>(ctxt, benefit, generatedNames), names(names),
+        symbolCache(symbolCache) {}
+
+protected:
+  Namespace &names;
+  SymbolHandler &symbolCache;
+};
+
+struct MemWritePortEnableAndMaskCanonicalizer
+    : public SymOpRewritePattern<MemoryWritePortOp> {
+  using SymOpRewritePattern::SymOpRewritePattern;
+  LogicalResult matchAndRewrite(MemoryWritePortOp op,
+                                PatternRewriter &rewriter) const final;
+};
+
+struct CallPassthroughArc : public SymOpRewritePattern<CallOp> {
+  using SymOpRewritePattern::SymOpRewritePattern;
+  LogicalResult matchAndRewrite(CallOp op,
+                                PatternRewriter &rewriter) const final;
+};
+
+struct StatePassthroughArc : public SymOpRewritePattern<StateOp> {
+  using SymOpRewritePattern::SymOpRewritePattern;
+  LogicalResult matchAndRewrite(StateOp op,
+                                PatternRewriter &rewriter) const final;
+};
+
+struct RemoveUnusedArcs : public SymOpRewritePattern<DefineOp> {
+  using SymOpRewritePattern::SymOpRewritePattern;
+  LogicalResult matchAndRewrite(DefineOp op,
+                                PatternRewriter &rewriter) const final;
+};
+
+} // namespace
+
+//===----------------------------------------------------------------------===//
+// Helpers
+//===----------------------------------------------------------------------===//
+
+LogicalResult canonicalizePassthoughCall(mlir::CallOpInterface callOp,
+                                         SymbolHandler &symbolCache,
+                                         PatternRewriter &rewriter) {
+  auto defOp = cast<DefineOp>(symbolCache.getDefinition(
+      callOp.getCallableForCallee().get<SymbolRefAttr>().getLeafReference()));
+  if (defOp.isPassthrough()) {
+    symbolCache.removeUser(defOp, callOp);
+    rewriter.replaceOp(callOp, callOp.getArgOperands());
+    return success();
+  }
+  return failure();
+}
+
+//===----------------------------------------------------------------------===//
+// Canonicalization pattern implementations
+//===----------------------------------------------------------------------===//
+
+LogicalResult MemWritePortEnableAndMaskCanonicalizer::matchAndRewrite(
+    MemoryWritePortOp op, PatternRewriter &rewriter) const {
+  auto defOp = cast<DefineOp>(symbolCache.getDefinition(op.getArcAttr()));
+  APInt enable;
+
+  if (op.getEnable() &&
+      mlir::matchPattern(
+          defOp.getBodyBlock().getTerminator()->getOperand(op.getEnableIdx()),
+          mlir::m_ConstantInt(&enable))) {
```

Comment:
That's a very neat way to look through the symbol and match against something in the arc definition!